src/mlpack/methods/kmeans/naive_kmeans_impl.hpp

```diff
@@ -29,75 +33,115 @@ NaiveKMeans<DistanceType, MatType>::NaiveKMeans(const MatType& dataset,
     distanceCalculations(0)
 { /* Nothing to do. */ }
 
-// Run a single iteration.
 template<typename DistanceType, typename MatType>
 double NaiveKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
                                                    arma::mat& newCentroids,
                                                    arma::Col<size_t>& counts)
 {
-  newCentroids.zeros(centroids.n_rows, centroids.n_cols);
-  counts.zeros(centroids.n_cols);
+  const size_t dims = dataset.n_rows;
+  const size_t points = dataset.n_cols;
+  const size_t clusters = centroids.n_cols;
+
+  newCentroids.zeros(dims, clusters);
+  counts.zeros(clusters);
 
-  // Find the closest centroid to each point and update the new centroids.
-  // Computed in parallel over the complete dataset
-  #pragma omp parallel
+  // Pre-compute squared norms of centroids
+  arma::vec centroidNorms(clusters);
+  #ifdef MLPACK_USE_OPENMP
+    #pragma omp parallel for schedule(static)
+  #endif
+  for (size_t j = 0; j < clusters; ++j)
   {
-    // The current state of the K-means is private for each thread
-    arma::mat localCentroids(centroids.n_rows, centroids.n_cols,
-        arma::fill::zeros);
-    arma::Col<size_t> localCounts(centroids.n_cols, arma::fill::zeros);
+    centroidNorms(j) = arma::dot(centroids.col(j), centroids.col(j));
+  }
+
+  // Determine the number of threads and calculate segment size
+  size_t effectiveThreads = 1;
+  #ifdef MLPACK_USE_OPENMP
+    const size_t numThreads = static_cast<size_t>(std::max(1, omp_get_max_threads()));
+    const size_t minVectorsPerThread = 100; 
+    effectiveThreads = std::min(numThreads, points / minVectorsPerThread);
+  #endif
+  const size_t nominalSegmentSize = points / effectiveThreads;
+
+  // Pre-allocate thread-local storage
+  std::vector<arma::mat> threadCentroids(effectiveThreads, arma::mat(dims, clusters, arma::fill::zeros));
+  std::vector<arma::Col<size_t>> threadCounts(effectiveThreads, arma::Col<size_t>(clusters, arma::fill::zeros));
```

Comment:
The `arma::fill::zeros` is redundant here, since the minimum version of Armadillo that mlpack requires now sets all internal elements to 0 when no fill type is specified in the constructor.  So you could simplify this part a bit 😄